Add "Publish your first course" task

fjorgemota opened the pull request:
Fixes #5639

### Changes proposed in this Pull Request

* Add Publish your First Course task;
* Add tests for the task added;
* Register task on the provider;

### Testing instructions

On a WP installation containing Sensei LMS on this branch:

1.  Run `curl --user "user:application password" http://your-host/wp-json/sensei-internal/v1/home` and make sure that the task "Publish your first Course" has `done` set to `false`, with the URL pointing to create a new course;
2.    Create a course/import the sample course, and verify if that task still has `done` set to `false`, with the URL still pointing to create a new course.
3. Publish the sample course, and verify if that task still has `done` set to `false`, with the URL still pointing to create a new course;
4. Now, create another course, and verify if that task still has `done` set to `false`, but with the URL now pointing to the edit page for that course you just created;
5. Publish the course you just created, and verify if that task now has `done` set to `true`, and the URL is now `null`;

fjorgemota committed:
Add "Publish your first Course" step

fjorgemota committed:
Register the new task on the tasks provider for Sensei Home

fjorgemota committed:
Fix tests for the Tasks Provider for Sensei Home

fjorgemota committed:
Move test for Create First Course task to task/ folder

fjorgemota committed:
Add tests for the Publish First Course task

fjorgemota commented:
I just rebased this branch over the changes on the `feature/new-onboarding-wizard`, and I fixed the conflict in that merge. I decided to rebase instead of merge because I assume no one downloaded this branch yet for review.

If you did, just delete your local branch and redownload again, please. :)

fjorgemota committed:
Remove get_image from the task "Publish your First Course"

jom commented on includes/admin/home/tasks/task/class-sensei-home-task-publish-first-course.php:
```diff
@@ -0,0 +1,94 @@
+			$prefix = $wpdb->esc_like( Sensei_Data_Port_Manager::SAMPLE_COURSE_SLUG );
+			// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery -- Safe-ish query.
+			$result = $wpdb->get_row( $wpdb->prepare( "SELECT ID, post_status='publish' as published FROM {$wpdb->posts} WHERE post_type='course' AND post_name NOT LIKE %s ORDER BY published DESC, ID DESC LIMIT 1", "{$prefix}%" ) );
+			wp_cache_set( $cache_key, $result, $cache_group );
```
I know we'd have to ignore the PHPCS warning, but I wonder if it is worth removing caching here. We're caching without any expiration. If there is any course (including trash), we're caching the object (without possibly `published === 0`) "forever". If there are no courses, `get_row` returns `null`. It is up to cache providers how to handle `null`, but it is possible that it will get stored and then `false !== $result` so it would also be returned "forever."

I think we also should probably do some filtering `post_status IN ('publish', 'draft')` so we never return trashed courses. Lastly, when all courses are in draft, do we want the latest created course or the _first_ course? If first, we might want to do `ID ASC`, I think.

fjorgemota replied:
Hey there!

> I know we'd have to ignore the PHPCS warning, but I wonder if it is worth removing caching here. We're caching without any expiration

As explained in https://github.com/Automattic/sensei/pull/5854#discussion_r995165448 - this shouldn't cache forever, given that I'm using `sensei/temporary` as cache group, and that cache group is added as non persistent (using this [WP function](https://developer.wordpress.org/reference/functions/wp_cache_add_non_persistent_groups/)) here: https://github.com/Automattic/sensei/blob/c50c2339b57308fa6f6e69a0892eedc361bdd4f1/includes/class-sensei.php#L350 

With that said:

> If there is any course (including trash)

As explained on the PR above, I created https://github.com/Automattic/sensei/issues/5934 to track this.

> we're caching the object (without possibly published === 0) "forever". If there are no courses, get_row returns null. It is up to cache providers how to handle null, but it is possible that it will get stored and then false !== $result so it would also be returned "forever."

I'm not sure if there's any effect to this given what I explained above. 

> Lastly, when all courses are in draft, do we want the latest created course or the first course? If first, we might want to do `ID ASC`, I think.

Good catch! I originally implemented `ID DESC` to look for the last course, as I assume it should be the very latest course the user worked on. However, now I'm not 100% sure. I feel like both should work well, but I think I'll change to `ID ASC` just to avoid surprises and to maintain some consistency.

fjorgemota replied:
Hey, fixed in [d758783](https://github.com/Automattic/sensei/pull/5888/commits/d75878341a70314a52f38a6e34133d239362b14d), [8f6ebb2](https://github.com/Automattic/sensei/pull/5888/commits/8f6ebb26faee378845af466e394bab8a3b0ab902) and 9e2f753938ab21e569546fc10536b47bcd7c2b32 

Hope it fixes all the points you mentioned, but let me know if it doesn't. :)

jom reviewed (changes requested):
Hi @fjorgemota! Looking good. I added one comment.

fjorgemota committed:
Ignore trashed posts on the task "Publish your First Course"

fjorgemota committed:
Always return the first course on the "Publish your First Course" task

fjorgemota committed:
Improve cache handling on the "Publish your first course" task

By not caching null values and also adding an expiration of 60 seconds for every value cached there.

jom reviewed (approved):
Looks great! 🚢
